### PR TITLE
#8630: support uint8 data type

### PR DESCRIPTION
**Pull request by hschoi4448**
### Ticket
- https://github.com/tenstorrent/tt-metal/issues/8630

### Problem description
- tt_eager library does not support the UINT8 data type
- Implementing the GenerateRandomBitMask operation requires using a bitmask tensor with the UINT8 data type.
- Currently, the tt_eager library does not support the UINT8 data type, making it impossible to create tensors of this type. The code snippet below illustrates this limitation.

```C++
// tt_eager/tensor/types.hpp
enum class DataType {
    BFLOAT16 = 0,
    FLOAT32 = 1,
    UINT32 = 2,
    BFLOAT8_B = 3,
    BFLOAT4_B = 4,
    UINT16 = 6,
    INT32 = 7,
    INVALID = 8,
};

# example to create UINT8 tt tensor
def test_uint8_tensor(device):
    npu_dtype = ttl.tensor.DataType.UINT8
    cpu_dtype = torch.uint8
    npu_layout = ttl.tensor.Layout.TILE
    input = torch.randint(0, 128, (32, 32), dtype=cpu_dtype)
    tt_input = ttl.tensor.Tensor(input, npu_dtype).to(npu_layout).to(device)
# E       AttributeError: type object 'tt_lib.tensor.DataType' has no attribute 'UINT8'
```

### What's changed
- support uint8 data type

### Checklist
- [x] Post commit CI passes (maybe?)
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by sangwon-chae** [comment]
@eyonland @arakhmati @adifrancescoTT Would you please review this PR?

**Comment by ayerofieiev-tt** [comment]
Screenshot 2024-06-21 at 12 45 42 AM

please going forward give a bit more context like why this is important. reviews not only help to make sure that code quality is good but also help to share knowledge. thank you!

**Review by arakhmati — APPROVED** [review]

**Comment by hschoi4448** [comment]
> Screenshot 2024-06-21 at 12 45 42 AM
> 
> please going forward give a bit more context like why this is important. reviews not only help to make sure that code quality is good but also help to share knowledge. thank you!

I have updated the problem description.

image

File: tt_eager/tensor/tensor_impl.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tensor/tensor_impl.cpp:251]: Should be `% 4`, right?

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_dm_ops.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_dm_ops.cpp:64]:
```suggestion
                Datatype must be one of the following types BFLOAT16, BFLOAT8_B, BFLOAT4_B, INT32, UINT32, UINT16 or UINT8.
```

**Review by ayerofieiev-tt — APPROVED** [review]
Looks good to me

File: tt_eager/tensor/tensor_impl.cpp
Comment (hschoi4448) [reply tt_eager/tensor/tensor_impl.cpp]: I have updated it to %4.

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_dm_ops.cpp
Comment (hschoi4448) [reply tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_dm_ops.cpp]:
```C++
Datatype must be one of the following types BFLOAT16, BFLOAT8_B, BFLOAT4_B, UINT32, INT32, UINT16 and UINT8.
```
As you suggested, I have made the changes.

**Comment by sangwon-chae** [comment]
@TT-BrianLiu @kmabeeTT @mywoodstock @tarafdarTT @tt-aho Please review this PR.

**Review by TT-BrianLiu — APPROVED** [review]
Comment about testing which should be addressed before merging

File: tests/ttnn/unit_tests/operations/test_unary.py
Comment (hschoi4448) [reply tests/ttnn/unit_tests/operations/test_unary.py]:
```Python
def comp_equal(golden, calculated):
   ...
   torch.equal(golden, calculated)
   ...

def assert_equal(...)
   ...
   comp_equal(...)
   ...  
```

I have added `comp_equal` and `assert_equal` to utility_functions to check for exact matches of integer values.


```Python
def run_identity_test(device, h, w, data_type, pcc=0.9999):
    torch.manual_seed(0)
    ... 

    if data_type == uint8:
        # init value
        torch_input_tensor = torch.randint(0, 245, (1, 1, h, w), dtype=torch.uint8)
        bias = 10

        ...
        assert_equal(torch_output_tensor, output_tensor)
    elif data_type == ttnn.uint16:
        # init value
        torch_input_tensor = torch.randint(0, 60000, (1, 1, h, w), dtype=torch.uint16)
        bias = 2000

        assert_equal(torch_output_tensor, output_tensor)
     ...
     else:
        # init value
        torch_input_tensor = torch.rand((1, 1, h, w), dtype=torch.bfloat16)
        
        assert_with_pcc(torch_output_tensor, output_tensor, pcc)

```
I have unrolled the tests for each `int` type and modified them to use `assert_equal` with `torch.equal` for int types.

**Comment by sangwon-chae** [comment]
@eyonland Would you please check this PR? Code owner review required. Thanks.

**Review by boris-drazic — APPROVED** [review]

**Comment by ayerofieiev-tt** [comment]
@tarafdarTT @kmabeeTT please take a look

File: tests/ttnn/unit_tests/operations/test_unary.py
Comment (TT-BrianLiu) [reply tests/ttnn/unit_tests/operations/test_unary.py]: looks good!

**Review by kmabeeTT — APPROVED** [review]

**Pull request merged**
